r-package/grf/vignettes/llf.Rmd

```diff
@@ -0,0 +1,414 @@
+
+```{r}
+# llf predictions from regression_forest
+results.llf <- predict(rforest, X.test, linear.correction.variables = 1:p) 
+preds.llf <- results.llf$predictions
+mean((preds.llf - truth)**2)
+
+# llf predictions from ll_regression_forest
+forest <- ll_regression_forest(X, Y, honesty = TRUE)
+results.llf <- predict(forest, X.test) 
+preds.llf <- results.llf$predictions
+mean((preds.llf - truth)**2)
+```
+
+### Weight Penalties 
+
+When we perform LLF predictions, we can either do a standard ridge regression (ll.weight.penalty set to FALSE), or scale by the covariance matrix (ll.weight.penalty set to TRUE): $\hat{\beta}_\text{TRUE} = (X'AX (1 + \lambda))^{-1} X'AY$. This defaults to FALSE.
+
+```{r}
+results.llf.unweighted <- predict(forest, X.test, 
+                               ll.weight.penalty = FALSE) 
+preds.llf.unweighted <- results.llf.unweighted$predictions
+mean((preds.llf.unweighted - truth)**2)
+
+results.llf.weighted <- predict(forest, X.test, 
+                               ll.weight.penalty = TRUE) 
+preds.llf.weighted <- results.llf.weighted$predictions
+mean((preds.llf.weighted - truth)**2)
+```
+
+### Residual Splits
+
+We also consider the role of tree training in local linear forests. A standard CART split minimizes prediction error from predicting leaf-wide averages. Instead, we can use residual splits, which minimize the corresponding prediction errors on ridge regression residuals. We might expect this to help in cases where there are strong linear signals from some covariates; we won't waste any forest splits modelling those, but can still discover them in the final regression step. Essentially this helps us make the most efficient possible splits in the forest, knowing that we have a local regression coming up to predict. This is currently an experimental feature.  
+
+```{r}
+forest <- ll_regression_forest(X, Y)
+preds.cart.splits <- predict(forest, X.test)
+
+ll.forest <- ll_regression_forest(X, Y, enable.ll.split = TRUE,
+                                  ll.split.weight.penalty = TRUE)
+preds.ll.splits <- predict(ll.forest, X.test)
+
+mse.cart.splits <- mean((preds.cart.splits$predictions - truth)^2)
+mse.ll.splits <- mean((preds.ll.splits$predictions - truth)^2)
+
+mse.cart.splits
+mse.ll.splits
+```
+
+To uncover exactly why this works, we can look at plots showing the split frequencies of both forests. In each plot, tiles represent how many splits at at given level (y-axis) of the tree were at each feature (x-axis). 
+
+```{r, message = FALSE}
+p <- 5
+XX <- matrix(runif(n * p, 0 ,1), nrow = n)
+YY <- apply(XX, MARGIN = 1, FUN = mu) + sigma * rnorm(n)
+
+forest2 <- regression_forest(XX, YY)
+
+max.depth <- 4
+freqs <- split_frequencies(forest2, max.depth = max.depth)
+d <- data.frame(freqs)
+dm <- melt(d)
```

Comment:
Instead of installing reshape2 on travis, could you reshape it in base R instead?
```
dm <- data.frame(variable = sort(rep(names(d), nrow(d))),
                 value = as.vector(as.matrix(d)))
```